Breaking change to await, update to flatten. (#2333)
dart-lang/language: specification/dartLangSpec.tex

which is used below and in other sections, as follows:

\begin{itemize}

> \item If $T$ is a type variable $X$ with bound $B$,

Reviewer: OK, so we're updating flatten anyway! But in that case we would basically have to assume the null safety specification as the context, because the new definition of flatten includes `S?` as a type. This is a little bit inconvenient because we don't have all the sections (that is, `\ref{X}` is needed for some `X` that don't exist), but I guess it should be OK to have a spec with unresolved refs for a while.

In any case, we'll land the null safety update any day now, of course! ;-)

Author: Of course, of course.

I decided to omit the `S?` case because I assumed it would be merged in with the rest of null safety,
so you only needed to merge the one line here into the patch.

Also why I didn't use `\DefEquals`, since the other existing cases didn't, and changing those would make rebasing the null safety PRs harder.

>   or $T$ is a promoted type variable $X & B$,

Author: Physical formatting (`\ ,\,`) ... ICK.
But what works works.

>   then $\flatten{T} = \flatten{B}$.

Reviewer: For consistency with other function definitions, we're using an equals sign with a triangle, defined by `\DefEquals{}{}`:
```suggestion
 then \DefEquals{\flatten{$T$}}{\code{\flatten{$B$}}}.
```

Author: The other cases don't use `\DefEquals`. I assumed the null safety PRs would make that change, and it was better to *not* change the other lines, so that rebasing got easier.

Author: Done.

> \item Otherwise if $T$ is \code{FutureOr<$S$>} for some $S$

Reviewer: Here's an updated version of the rest of the definition of flatten (replacing lines 11274--11280, including both):

```
\item If $T$ is \code{$S$?}\ for some $S$
  then \DefEquals{\flatten{$T$}}{\code{\flatten{$S$}?}}.
\item If $T$ is \code{FutureOr<$S$>} then \DefEquals{\flatten{$T$}}{S}.
\item If $T$ implements \code{Future<$S$>}
  (\ref{interfaceSuperinterfaces})
  then \DefEquals{\flatten{$T$}}{S}.
\item Otherwise,
  \DefEquals{\flatten{$T$}}{T}.
```

Reviewer: We'd need to add the following to `dart.sty`, (preferably, just before `% ......` at the end, such that it will merge cleanly with the dart.sty from the null safety update):

```
\newcommand{\DefEquals}[2]{\ensuremath{{#1}\stackrel{\vartriangle}{=}{#2}}}
\newcommand{\DefEqualsNewline}[2]{
  \ensuremath{{#1}\stackrel{\vartriangle}{=}}\\
  \ensuremath{{#2}}%
}
```

}

\commentary{%

> An member access on a type literal

Reviewer: ```suggestion
A member access on a type literal
```

% NOTICE: Removed the requirement that an error thrown by $e$ is caught in a
% future. There is no reason $var x = e; await x;$ and $await e$ should behave
% differently, and no implementation actually implemented it.
Let $R$ be the run-time type of $o$.
If $R$ implements \code{Future<$S$>} for some $S$
(\ref{interfaceSuperinterfaces}) then
it is a dynamic type error if $S$ is not a subtype of $T$.

> If $S$ is a subtype of $T$, let $f$ be $o$.

Reviewer: I usually handle descriptions of errors (run-time or compile-time) slightly differently:
```suggestion
Otherwise
(\commentary{if no run-time error occurred}),
let $f$ be $o$.
```

We could do that for consistency. However, I also think this style is a bit safer to read (a bit harder to misunderstand), because it puts the emphasis on the error: "One case is an error, all other cases proceed as follows".

With 'If $S$ is a subtype of $T$', there's more room to think that there could be even more cases.

> Otherwise (when $R$ does not implement \code{Future}),

Reviewer: ```suggestion
Otherwise
(\commentary{when $R$ does not implement \code{Future}}),
```

> let $f$ be a future object, implementing \code{Future<$T$>},

Reviewer: ```suggestion
let $f$ be an object whose run-time type implements \code{Future<$T$>},
```

% NOTICE: Removed the requirement that an error thrown by $e$ is caught in a
% future. There is no reason $var x = e; await x;$ and $await e$ should behave
% differently, and no implementation actually implemented it.
Let $R$ be the run-time type of $o$.
If $R$ implements \code{Future<$S$>} for some $S$
(\ref{interfaceSuperinterfaces}) then
it is a dynamic type error if $S$ is not a subtype of $T$.
Otherwise
(\commentary{if no run-time error occurred}),
let $f$ be $o$.

Otherwise
(\commentary{when $R$ does not implement \code{Future}}),
let $f$ be an object whose run-time type implements \code{Future<$T$>},

> which is completed with the value $o$.

Author: Nit. This doesn't actually say what behavior the *f* object has.
